src/OpenTelemetry.Instrumentation.Dependencies/Implementation/GrpcClientDiagnosticListener.cs

```diff
@@ -0,0 +1,102 @@
+using System.Reflection;
+using System.Text.RegularExpressions;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.Dependencies.Implementation
+{
+    internal class GrpcClientDiagnosticListener : ListenerHandler
+    {
+        private static readonly Regex GrpcMethodRegex = new Regex(@"(?<service>\w+)/(?<method>\w+)", RegexOptions.Compiled);
+
+        private readonly ActivitySourceAdapter activitySource;
+        private readonly PropertyFetcher startRequestFetcher = new PropertyFetcher("Request");
+        private readonly PropertyInfo activityKindPropertyInfo = typeof(Activity).GetProperty("Kind");
+
+        public GrpcClientDiagnosticListener(ActivitySourceAdapter activitySource)
+            : base("Grpc.Net.Client", null)
+        {
+            if (activitySource == null)
+            {
+                throw new ArgumentNullException(nameof(activitySource));
+            }
+
+            this.activitySource = activitySource;
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStartActivity";
+            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
+            {
+                InstrumentationEventSource.Log.NullPayload(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                return;
+            }
+
+            var grpcMethodTag = activity.Tags.FirstOrDefault(tag => tag.Key == "grpc.method");
+            var grpcMethod = grpcMethodTag.Value?.Trim('/');
+
+            // TODO: Avoid the reflection hack once .NET ships new Activity with Kind settable.
+            this.activityKindPropertyInfo.SetValue(activity, ActivityKind.Client);
+            activity.DisplayName = grpcMethod;
+
+            this.activitySource.Start(activity);
+
+            if (activity.IsAllDataRequested)
+            {
+                activity.AddTag("rpc.system", "grpc");
+                var rpcService = GrpcMethodRegex.Match(grpcMethod).Groups["service"].Value;
+
+                activity.AddTag("rpc.service", rpcService);
+
+                var uriHostNameType = Uri.CheckHostName(request.RequestUri.Host);
+                if (uriHostNameType == UriHostNameType.IPv4 || uriHostNameType == UriHostNameType.IPv6)
+                {
+                    activity.AddTag("net.peer.ip", request.RequestUri.Host);
+                }
+                else
+                {
+                    activity.AddTag("net.peer.name", request.RequestUri.Host);
+                }
+
+                activity.AddTag("net.peer.port", request.RequestUri.Port.ToString());
+            }
+        }
+
+        public override void OnStopActivity(Activity activity, object payload)
+        {
+            if (activity.IsAllDataRequested)
+            {
+                var status = Status.Unknown;
+
+                var grpcStatusCodeTag = activity.Tags.FirstOrDefault(tag => tag.Key == "grpc.status_code").Value;
```

Comment:
can you define a constant for those tag names populated by the grpcClient? "grpc.method" too I believe.

It'd be good to comment that this tag is populated the actual grpcClient library. For most other instrumentation, the library itself doesn't populate any tags at all.

Reply:
Good idea. Constants added.